tactics/proof_global.ml

```diff
@@ -164,31 +164,20 @@ let return_proof { proof } =
      equations and so far there is no code in the CI that will
      actually call those and do a side-effect, TTBOMK *)
   (* EJGA: likely the right solution is to attach side effects to the first constant only? *)
-  let proofs = List.map (fun (c, _) -> (proof_opt c, eff)) initial_goals in
+  let proofs = List.map (fun (body, typ) -> (proof_opt body, eff), proof_opt typ) initial_goals in
   proofs, Evd.evar_universe_context evd
 
 let close_proof ~opaque ~keep_body_ucst_separate ps =
-  let elist, uctx = return_proof ps in
+  let elist, uctx = prepare_proof ps in
   let { section_vars; proof; udecl; initial_euctx } = ps in
-  let { Proof.name; poly; entry; sigma } = Proof.data proof in
+  let { Proof.name; poly } = Proof.data proof in
   let opaque = match opaque with Opaque -> true | Transparent -> false in
 
-  (* Because of dependent subgoals at the beginning of proofs, we could
-     have existential variables in the initial types of goals, we need to
-     normalise them for the kernel. *)
-  let subst_evar k = Evd.existential_opt_value0 sigma k in
-  let nf = UnivSubst.nf_evars_and_universes_opt_subst subst_evar (UState.subst uctx) in
-
-  let make_entry (body, eff) (_, typ) =
+  let make_entry ((body, eff), typ) =
     let allow_deferred =
       not poly && (keep_body_ucst_separate ||
                    not (Safe_typing.empty_private_constants = eff.Evd.seff_private))
     in
-    (* EJGA: Why are we doing things this way? *)
-    let typ = EConstr.Unsafe.to_constr typ in
-    let typ = if allow_deferred then typ else nf typ in
```

Comment:
Removing this special-case seems like the main candidate to break `validate: quick`; that's extremely fishy, we'll have to see what is going on the whole `quick` stuff.

Reply:
Yeah it's critical, IIRC what happens is the type involves some universe `u`, and during the proof it gets assigned `u := v`. Then in validate you check that you got the expected type but there's a missing universe constraint or something (I don't remember th details).

Reply:
Thanks for the info @SkySkimmer

Reply:
Ok so with the super-horrible hack I added to prepare things work; but I feel like they shouldn't.

Reply:
The hack is not so horrible now I think.